posts/inside-rust/2020-03-19-terminating-rust.md

```diff
@@ -0,0 +1,51 @@
+---
+layout: post
+title: "Resolving Rust's forward progress guarantees"
+author: Mark Rousskov
+description: "Should side-effect be the fix?"
+team: the compiler team <https://www.rust-lang.org/governance/teams/compiler>
+---
+
+There has been a longstanding miscompilation in Rust: programs that do not make
+[forward progress]. Note that the previous link is to the C++ definition; Rust
+is not C++, but currently LLVM optimizes all LLVM IR with the assumption that a
+lack of forward progress is undefined behavior.
+
+Note also that Rust does not define a lack of forward progress as [undefined
+behavior], while C++ does. It is particularly common to encounter the
+miscompilation "intentionally" when writing panic handlers and other such code
+with a body of `loop {}`. Some users also report that they've unintentionally
+hit this bug in recursive code which accidentally lacks a base case.
+
+Somewhat recently, LLVM added an intrinsic which tells the optimizer that
+forward progress has been made. On nightly Rust, you can enable this with
+`-Zinsert-sideeffect`.
```

Comment:
maybe give a bit more detail here -- i.e., "this will cause the compiler to automatically insert calls to the intrinsic in places that LLVM might consider to be infinite loops." or something.